README.md
# solitaire-encryption

Implements the [Solitaire encryption algorithm](https://www.schneier.com/academic/solitaire/) by Bruce Schneier and featured in the novel Cryptonomicon by Neal Stephenson.

Note that this is a completely silly thing to do, since the point of the algorithm is to be able to use it by hand with a physical deck of cards. This is just a Haskell exercise `:)`


CHANGELOG.md
# Changelog for `solitaire-encryption`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


test/KeySpec.hs
{-# OPTIONS_GHC -Wno-unused-imports #-}

module KeySpec (main, spec) where

import Test.Hspec
import Test.QuickCheck

import Deck
import Key

-- `main` is here so that this module can be run from GHCi on its own.  It is
-- not needed for automatic spec discovery.
main :: IO ()
main = hspec spec

spec :: Spec
spec = return ()


test/DeckSpec.hs
{-# OPTIONS_GHC -Wno-unused-imports #-}

module DeckSpec (main, spec) where

import Test.Hspec
import Test.QuickCheck

import Deck

-- `main` is here so that this module can be run from GHCi on its own.  It is
-- not needed for automatic spec discovery.
main :: IO ()
main = hspec spec

spec :: Spec
spec = do
  describe "cardValue" $ do
    it "correctly values the ordered deck" $ do
      cardValue <$> orderedDeck `shouldBe` [1 .. 53] ++ [53]

  describe "moveCard" $ do
    it "correctly moves a card down" $ moveCard (spade Two) 2 spadesOneThroughFive
      `shouldBe` spades [Ace, Three, Four, Two, Five]
    it "correctly wraps a card around" $ moveCard (spade Four) 2 spadesOneThroughFive
      `shouldBe` spades [Ace, Four, Two, Three, Five]

  describe "jokersTripleCut" $ do
    it "should do a triple cut" $ jokersTripleCut (insertJokers (spades [Ace, Two]) (spades [Three, Four]) (spades [Five, Six]))
      `shouldBe` insertJokers (spades [Five, Six]) (spades [Three, Four]) (spades [Ace, Two])
    it "should not depend on joker order" $ jokersTripleCut (insertJokers' (spades [Ace, Two]) (spades [Three, Four]) (spades [Five, Six]))
      `shouldBe` insertJokers' (spades [Five, Six]) (spades [Three, Four]) (spades [Ace, Two])
    it "should handle empty left" $ jokersTripleCut (insertJokers [] (spades [Three, Four]) (spades [Five, Six]))
      `shouldBe` insertJokers (spades [Five, Six]) (spades [Three, Four]) []
    it "should handle empty right" $ jokersTripleCut (insertJokers (spades [Ace, Two]) (spades [Three, Four]) [])
      `shouldBe` insertJokers [] (spades [Three, Four]) (spades [Ace, Two])
    it "should handle empty middle" $ jokersTripleCut (insertJokers (spades [Ace, Two]) [] (spades [Five, Six]))
      `shouldBe` insertJokers (spades [Five, Six]) [] (spades [Ace, Two])
    it "should handle empty ends" $ jokersTripleCut (insertJokers [] (spades [Three, Four]) [])
      `shouldBe` insertJokers [] (spades [Three, Four]) []

  describe "countCut" $ do
    it "should do the thing" $ do
      countCut (clubs [Four, Five, Ace, Two, Three]) `shouldBe` clubs [Two, Four, Five, Ace, Three]

  describe "countCard" $ do
    it "should work" $ do
      countCard (clubs [Four, Five, Ace, Two, Three]) `shouldBe` Card Club Two

spade :: Rank -> Card
spade = Card Spade

spades :: [Rank] -> Deck
spades = map (Card Spade)

clubs :: [Rank] -> Deck
clubs = map (Card Club)

spadesOneThroughFive :: Deck
spadesOneThroughFive = Card Spade <$> [Ace, Two, Three, Four, Five]

insertJokers :: Deck -> Deck -> Deck -> Deck
insertJokers l c r = l ++ JokerA : c ++ JokerB : r

insertJokers' :: Deck -> Deck -> Deck -> Deck
insertJokers' l c r = l ++ JokerB : c ++ JokerA : r

test/SolitaireSpec.hs
{-# OPTIONS_GHC -Wno-unused-imports #-}

module SolitaireSpec (main, spec) where

import Test.Hspec
import Test.QuickCheck

import Solitaire

-- `main` is here so that this module can be run from GHCi on its own.  It is
-- not needed for automatic spec discovery.
main :: IO ()
main = hspec spec

spec :: Spec
spec = do
    describe "encode" $ do
        it "encodes a string" $ do
            encode "Do Not Use PC!" `shouldBe` [4, 15, 14, 15, 20, 21, 19, 5, 16, 3]

    describe "ltr" $ do
        it "does the thing" $ do
            ltr <$> [1, 5, 13, 26] `shouldBe` "AEMZ"

    describe "addLetters" $ do
        it "adds A + E = F" $ do
            pos 'A' +* pos 'E' `shouldBe` pos 'F'
        it "adds T + Q = K" $ do
            pos 'T' +* pos 'Q' `shouldBe` pos 'K'

    describe "subLetters" $ do
        it "subtracts F - A = E" $ do
            pos 'F' -* pos 'A' `shouldBe` pos 'E'
        it "subtracts K - Q = T" $ do
            pos 'K' -* pos 'Q' `shouldBe` pos 'T'



src/Shuffle.hs
{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}

module Shuffle (shuffledDeck) where

import Deck
import System.Random (RandomGen, randomR)

shuffledDeck :: RandomGen g => g -> ([Card], g)
shuffledDeck g = shuffleList g orderedDeck

-- Shuffles an entire list (must be finite) with the given RandomGen, and gives back a new generator
shuffleList :: RandomGen g => g -> [a] -> ([a], g)
shuffleList g xs = shuffleList' g (length xs) xs

-- Shuffles the first n elements of a list with the given RandomGen, and gives back a new generator.
-- The list need not be finite, but it does need to have at least length n.
-- Note also that if the generator isn't actually used, the same one is returned
-- This algorithm is probably pretty inefficient, relying on many list concats
-- ...but since the list isn't really that big, probably not a big deal
shuffleList' :: RandomGen g => g -> Int -> [a] -> ([a], g)
shuffleList' _ n _ | n < 0 = error "shuffleList' not defined for negative n"
shuffleList' g 0 xs = (xs, g)
shuffleList' g 1 xs = (xs, g)
shuffleList' g n xs = (x : shuffledRest, g'') where
    (i, g') = randomR (0, n - 1) g
    (xsL, x : xsR) = splitAt i xs
    (shuffledRest, g'') = shuffleList' g' (n - 1) (xsL ++ xsR)


src/Solitaire.hs
module Solitaire (encrypt, decrypt, encode, keystream, pos, ltr, (|>), (+*), (-*)) where

import Deck
import Data.Char (toUpper, ord, chr, isAsciiUpper, isAsciiLower)

encrypt :: String -> Deck -> (String, Deck)
encrypt s d = let
    clear = encode s
    (ks, d') = keystream (length clear) d
    cipher = ltr <$> zipWith (+*) clear ks
    in (cipher, d')

decrypt :: String -> Deck -> (String, Deck)
decrypt s d = let
    cipher = pos <$> s
    (ks, d') = keystream (length cipher) d
    clear = ltr <$> zipWith (-*) cipher ks
    in (clear, d')

-- reverse composition for easier reading left to right
(|>) :: (a -> b) -> (b -> c) -> (a -> c)
f |> g = g . f

-- generates n keystream values from the given deck. the deck should be a full 54-card deck, including
-- both jokers; this will break if either joker is missing, and it will generate nonsense if
-- any of the non-Joker cards are missing or duplicated.
keystream :: Int -> Deck -> ([Int], Deck)
keystream 0 d = ([], d)
keystream n d = let
    steps = moveCard JokerA 1 |> moveCard JokerB 2 |> jokersTripleCut |> countCut
    d' = steps d
    outCard = countCard d'
    in if isJoker outCard
        then keystream n d' -- go again without generating output for this step
        else let
            out = cardValue' outCard
            (outTail, d'') = keystream (n - 1) d'
            in (out : outTail, d'')


-- converts a string into a list of ints, by throwing away non-alphabetic characters and
-- converting letters to their positional values (A = 1, B = 2, ..., Z = 26), ignoring case
encode :: String -> [Int]
encode "" = []
encode (c:cs)
    | isAsciiUpper c = pos c : encode cs
    | isAsciiLower c = pos (toUpper c) : encode cs
    | otherwise = encode cs -- throw away this character

-- converts an uppercase letter to its positional value: A = 1, B = 2, ..., Z = 26
-- produces nonsense for characters that aren't uppercase letters from the English alphabet (ascii 65 thru 90)
pos :: Char -> Int
pos c = 1 + (ord c - ord 'A')

ltr :: Int -> Char
ltr i = chr (i + ord 'A' - 1)

-- adds two numbers that are assumed to correspond to letters (i.e. be between 1 and 26 inclusive),
-- wrapping around so that e.g. 'A + C = D' and 'Z + C = C'
(+*) :: Int -> Int -> Int
a +* b
    | a + b > 26 = a + b - 26
    | otherwise  = a + b

-- same thing but in reverse
(-*) :: Int -> Int -> Int
a -* b
    | a > b = a - b
    | otherwise = 26 + a - b


src/Deck.hs
{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}

module Deck (
    Suit(..), Rank(..), Card(..), Deck, orderedDeck, cardValue, isJoker,
    moveCard, insertCardAt, jokersTripleCut, countCut, countCard, cardValue',
    countCut', jokers, suit
) where

data Suit = Club | Diamond | Heart | Spade deriving (Eq, Ord)

instance Show Suit where
    show Club       = "♣"
    show Diamond    = "♦"
    show Heart      = "♥"
    show Spade      = "♠"

suit :: Suit -> Deck
suit s = Card s <$> ranks

suits :: [Suit]
suits = [Club, Diamond, Heart, Spade]

data Rank = Ace | Two | Three | Four | Five | Six | Seven | Eight | Nine | Ten | Jack | Queen | King deriving (Eq, Ord)

instance Show Rank where
    show Ace    = "A"
    show Two    = "2"
    show Three  = "3"
    show Four   = "4"
    show Five   = "5"
    show Six    = "6"
    show Seven  = "7"
    show Eight  = "8"
    show Nine   = "9"
    show Ten    = "10"
    show Jack   = "J"
    show Queen  = "Q"
    show King   = "K"

ranks :: [Rank]
ranks = [Ace, Two, Three, Four, Five, Six, Seven, Eight, Nine, Ten, Jack, Queen, King]

data Card = Card Suit Rank | JokerA | JokerB deriving (Eq, Ord)

instance Show Card where
    show (Card s r) = show r ++ show s
    show JokerA     = "JokerA"
    show JokerB     = "JokerB"

type Deck = [Card]

jokers :: Deck
jokers = [JokerA, JokerB]

isJoker :: Card -> Bool
isJoker (Card _ _) = False
isJoker _ = True

orderedDeck :: Deck
orderedDeck = foldr (\s d -> suit s ++ d) jokers suits

cardValue :: Card -> Int
cardValue (Card s r)    = suitValue s + rankValue r
cardValue JokerA        = 53
cardValue JokerB        = 53 -- both jokers have the same value

-- Clubs/Hearts are 1--13, Spades/Diamonds are 14--26
cardValue' :: Card -> Int
cardValue' (Card Heart r) = cardValue (Card Club r)
cardValue' (Card Spade r) = cardValue (Card Diamond r)
cardValue' c = cardValue c

suitValue :: Suit -> Int
suitValue Club      = 0
suitValue Diamond   = 13
suitValue Heart     = 26
suitValue Spade     = 39

rankValue :: Rank -> Int
rankValue Ace   = 1
rankValue Two   = 2
rankValue Three = 3
rankValue Four  = 4
rankValue Five  = 5
rankValue Six   = 6
rankValue Seven = 7
rankValue Eight = 8
rankValue Nine  = 9
rankValue Ten   = 10
rankValue Jack  = 11
rankValue Queen = 12
rankValue King  = 13

-- Deck operations relevant to the Solitaire algorithm

-- moves the (first instance of the) given card down the deck the given number of spaces, wrapping around to the top of the deck
-- if the bottom is reached. The given card is assumed to exist in the deck, and the number of spaces to move the card must be
-- less than the size of the deck
moveCard :: Card -> Int -> Deck -> Deck
moveCard c n d = let
    (l, r) = break (== c) d
    r' = tail r
    in if n <= length r'
        then l ++ insertCardAt c n r' -- insert n cards down towards the end
        else insertCardAt c (n - length r') l ++ r' -- wrap around from the beginning

-- inserts a card at the given position in the deck; `insertCardAt 0 2 [1, 2, 3, 4] = [1, 2, 0, 3, 4]`
insertCardAt :: Card -> Int -> Deck -> Deck
insertCardAt c n d = let (l, r) = splitAt n d in l ++ c : r

-- a deck like (cards 1) Joker (cards 2) Joker (cards 3) becomes (cards 3) Joker (cards 2) Joker (cards 1)
-- the set of cards at the beginning or end can be empty; if both are empty, the deck is unchanged.
-- Two jokers must be in the deck.
jokersTripleCut :: Deck -> Deck
jokersTripleCut d = let
    (l, jcjr) = break isJoker d -- Left and Joker1|Center|Joker2|Right
    j1 : cjr = jcjr
    (c, jr) = break isJoker cjr
    j2 : r = jr

    in r ++ (j1 : c) ++ (j2 : l)

-- cuts the deck at the location indicated by the last card's value (1--53, see `cardValue` above),
-- leaving the last card in place. So if the last card's value is 3, a deck in order (1 2 3 ... 52 53 54)
-- becomes (4 5 6 ... 52 53 1 2 3 54).
-- Deck must have all 54 cards - or, more precisely, it must have at least as many cards as the maximum cardValue present.
countCut :: Deck -> Deck
countCut d = let
    i = cardValue (last d)
    in countCut' i d

countCut' :: Int -> Deck -> Deck
countCut' i d = let
    (l, r) = splitAt i d
    in init r ++ l ++ [last r]

-- for a given deck, gets the value x of the top card, then gets the value of the (x + 1)th card down from the top.
countCard :: Deck -> Card
countCard d = let
    i = cardValue (head d)
    in d !! (i - 1)
